compiler/rustc_hir_typeck/src/coercion.rs

```diff
@@ -932,26 +933,38 @@ impl<'f, 'tcx> Coerce<'f, 'tcx> {
                         return Err(TypeError::ForceInlineCast);
                     }
 
-                    let fn_attrs = self.tcx.codegen_fn_attrs(def_id);
-                    if matches!(fn_attrs.inline, InlineAttr::Force { .. }) {
-                        return Err(TypeError::ForceInlineCast);
-                    }
-
-                    // FIXME(target_feature): Safe `#[target_feature]` functions could be cast to safe fn pointers (RFC 2396),
-                    // as you can already write that "cast" in user code by wrapping a target_feature fn call in a closure,
-                    // which is safe. This is sound because you already need to be executing code that is satisfying the target
-                    // feature constraints..
                     if b_hdr.safety.is_safe()
                         && self.tcx.codegen_fn_attrs(def_id).safe_target_features
                     {
-                        return Err(TypeError::TargetFeatureCast(def_id));
+                        // Allow the coercion if the current function has all the features that would be
+                        // needed to call the coercee safely.
+                        let coercee_features = &self.tcx.codegen_fn_attrs(def_id).target_features;
+                        let body_features =
+                            &self.tcx.codegen_fn_attrs(self.fcx.body_id).target_features;
+                        if !self.tcx.sess.target.options.is_like_wasm
+                            && !coercee_features
+                                .iter()
+                                .all(|feature| body_features.iter().any(|f| f.name == feature.name))
+                        {
+                            return Err(TypeError::TargetFeatureCast(def_id));
+                        } else {
+                            allow_unsafe_to_safe_coercion = true;
```

Comment:
you could just mutate the `a_sig` binding directly here instead of mutating a bool and using that to later mutate `a_sig`.